File: neetcode/blind75_rust/src/code/container_with_most_water.rs
use std::cmp;

pub struct Solution {}

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut ret = 0;
        let mut i = 0;
        let mut j = height.len() - 1;

        while i < j {
            let area = (j - i) as i32 * cmp::min(height[j], height[i]);
            ret = cmp::max(ret, area);

            if height[i] < height[j] {
                i += 1;
            } else {
                j -= 1;
            }
        }

        return ret as i32;
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn case_1() {
        let height = vec![1, 8, 6, 2, 5, 4, 8, 3, 7];
        let expected = 49;
        let actual = Solution::max_area(height);

        assert_eq!(actual, expected);

        let height2 = vec![1, 1];
        let expected2 = 1;
        let actual2 = Solution::max_area(height2);

        assert_eq!(actual2, expected2);
    }
}

File: python/container_with_most_water.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        """time complexity: O(n), space complexity: O(1)"""
        result: int = 0
        l: int = 0
        r: int = len(heights) - 1
        while l < r:
            area: int = min(heights[l], heights[r]) * (r - l)
            result = max(result, area)
            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1

        return result


def test_maxArea_case_1():
    # arrange
    height: list[int] = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    expected: int = 49

    # act
    solution = Solution()
    actual = solution.maxArea(height)

    # assert
    assert expected == actual


def test_maxArea_case_2():
    # arrange
    height: list[int] = [1, 1]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.maxArea(height)

    # assert
    assert expected == actual


def test_maxArea_case_3():
    # arrange
    height: list[int] = [2, 3, 4, 5, 18, 17, 6]

    expected: int = 17

    # act
    solution = Solution()
    actual = solution.maxArea(height)

    # assert
    assert expected == actual
